README.md

```diff
@@ -90,9 +158,169 @@ reproducible.
+
+- A reproducible, CI-ready Check (run it with `garn check frontend.test`)
+- A no-installation-needed Executable (run it with `garn check frontend.test`).
+- An Environment or devshell you can enter with `garn enter frontend` with all
+  dependencies available.
+
+The right version of `node`, `npm`, `jq`, etc. will be used for all these.
+You can see a full project [here](https://github.com/garnix-io/garn/tree/main/examples/npm-project).
+
+### Go
+
+Go projects are usually built with `mkGoProject`:
+
+```typescript
+export const server = garn.go.mkGoProject({
+    description: "example backend server in go",
+    src: ".",
+    goVersion: "1.21",
+  })
+  .addExecutable("migrate", "go run ./scripts/migrate.go")
+  .addExecutable("dev", "go run ./main.go");
+```
+
+### Haskell
+
+For creating Haskell projects, there's `mkHaskellProject`. An example:
+
+```typescript
+export const project = garn.haskell.mkHaskellProject({
+  description: "My project",
+  src: ".",
+  executables: ["server", "openapi-generation"]
+  ghcVersion: "ghc94"
+})
+```
+
+Assuming you have a `cabal` file in `.`, this creates a Project containing:
+
+- A reproducible, CI-ready Package (run it with `garn build project`) that
+  includes the test-suites of your project.
+- Two executables, that run the corresponding executables from the cabal file.
+  Run them with `garn run project.server` and `garn run project.openapi-generation`.
+- An Environment or devshell you can enter with `garn enter project` with all
+  dependencies available.
+
+The right version of `node`, `npm`, `jq`, etc. will be used for all these.
+You can see a full project [here](https://github.com/garnix-io/garn/tree/main/examples/npm-project).
+
+## How it works
+
+The essential idea is to generate Nix code corresponding to your projects, and
+then run Nix behind the scenes. In [here](https://github.com/garnix-io/garn/blob/main/ts/nix.ts),
+you can see the simple but low-level Nix AST we use. Objects such as `Check`s
+have a `nixExpression` field that contains their corresponding Nix expression.
+
+**garn** then generates a `flake.nix` that, for every exported variables of the `garn.ts`,
+has a corresponding `package`, `check`, `devshell` or `app`.
+
+This means that **garn** integrates with Nix (and the Nix ecosystem) *both ways*:
+you can write pure Nix expressions in **garn**, and import Nix projects, but
+you can also from Nix call things generated by **garn**.
 
 ## Typescript API
 
-Documentation for the `garn` Deno library is documented
-[here](https://doc.deno.land/https://garn.io/ts/v0.0.18/mod.ts).
+You can find documentation for the **garn** Deno library [here](https://doc.deno.land/https://garn.io/ts/v0.0.18/mod.ts).
+
+
+## Comparison to other tools
+
+- **Nix**: Nix is a brilliant idea with an immense amount of volunteer effort
+  behind it, and the giant shoulders on which **garn** stands. It has a rather
+  complex CLI, and the Nix language is a whole new thing to learn. That said
```

Comment:
Maybe mention that nix is also not very accessible and/or has a very steep learning curve?